### PR TITLE
support tls

### PR DESCRIPTION
**Pull request by marsishandsome**
Signed-off-by: marsishandsome <marsishandsome@gmail.com>

### What problem does this PR solve? 
https://github.com/tikv/client-java/issues/267

### What is changed and how it works?
support tls

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: src/main/java/org/tikv/common/util/ChannelFactory.java [hunk 1]
```diff
@@ -53,14 +94,32 @@ public ManagedChannel getChannel(String addressStr, HostMapping hostMapping) {
...
+              return null;
+            }
+            return NettyChannelBuilder.forAddress(mappedAddr.getHost(), mappedAddr.getPort())
+                .maxInboundMessageSize(maxFrameSize)
+                .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
+                .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
+                .keepAliveWithoutCalls(true)
+                .sslContext(sslContext)
+                .idleTimeout(60, TimeUnit.SECONDS)
+                .build();
```
Comment (andylokandy) [inline src/main/java/org/tikv/common/util/ChannelFactory.java:121]: Is it possible to DRY?

Comment (marsishandsome) [reply src/main/java/org/tikv/common/util/ChannelFactory.java:121]: not support yet

Comment (andylokandy) [reply src/main/java/org/tikv/common/util/ChannelFactory.java:121]: I think it is possible, see the example 11 in https://www.programcreek.com/java-api-examples/?class=io.grpc.netty.NettyChannelBuilder&method=sslContext

Comment (marsishandsome) [reply src/main/java/org/tikv/common/util/ChannelFactory.java]: fixed  @andylokandy PTAL

**Review by andylokandy — APPROVED** [review]

File: src/main/java/org/tikv/common/TiConfiguration.java [hunk 2]
```diff
@@ -689,4 +695,36 @@ public int getKeepaliveTimeout() {
...
+
+  public void setTlsEnable(boolean tlsEnable) {
+    this.tlsEnable = tlsEnable;
+  }
+
+  public Optional<String> getTrustCertCollectionFile() {
+    return trustCertCollectionFile;
+  }
+
+  public void setTrustCertCollectionFile(Optional<String> trustCertCollectionFile) {
```
Comment (kolbe) [inline src/main/java/org/tikv/common/TiConfiguration.java:711]: Is it normal to use Optional<T> as a method argument? It seems like this is discouraged: https://rules.sonarsource.com/java/RSPEC-3553

Comment (marsishandsome) [reply src/main/java/org/tikv/common/TiConfiguration.java]: fixed, PTAL

File: README.md [hunk 3]
```diff
@@ -152,6 +152,24 @@ The following includes ThreadPool related parameters, which can be passed in thr
 - whether to enable `Compare And Set`, set true if using `RawKVClient.compareAndSet` or `RawKVClient.putIfAbsent`
 - default: false
 
+### TLS
+
+#### tikv.tls_enable
+- whether to enable tls
```
Comment (dveeden) [inline README.md:158]:
```suggestion
- whether to enable TLS
```

File: README.md [hunk 4]
```diff
@@ -152,6 +152,24 @@ The following includes ThreadPool related parameters, which can be passed in thr
...
+- default: false
+
+#### tikv.trust_cert_collection
+- Trusted certificates for verifying the remote endpoint's certificate, e.g. /home/tidb/ca.pem. The file should contain an X.509 certificate collection in PEM format.
```
Comment (dveeden) [inline README.md:162]: The name of this setting looks like something that could be a boolean. But this is the path to the `ca.pem` file?

We *could* consider to set this to `/etc/pki/tls/cert.pem` by default.

```
$ openssl version -d
OPENSSLDIR: "/etc/pki/tls"
```

**Review by dveeden — COMMENTED** [review]
Is there anything we need to do now (or later as follow-up) for:

- Restricting this to TLSv1.1 or newer?
- Removing insecure cipher suites (like 3DES)?

File: README.md (hunk 4 above)
Comment (marsishandsome) [reply README.md:162]: default is null now.

Comment (marsishandsome) [reply README.md:162]: Users can pass the argument by themself.

**Pull request merged**